pkg/controller/sparkapplication/controller.go

```diff
@@ -607,37 +602,55 @@ func (c *Controller) recordExecutorEvent(
 	}
 }
 
-// shouldSubmit determines if a given application is subject to a submission retry.
+// shouldSubmit determines if a given application is subject to a retry.
 func shouldSubmit(app *v1alpha1.SparkApplication) bool {
-	// App was never attempted. Should submit.
-	if app.Status.Attempts == 0 {
-		return true
-	}
 
-	// No retries specified or all retries exhaused.
-	if app.Spec.FailureRetries == nil ||
-		app.Status.Attempts >= (*app.Spec.FailureRetries)+1 {
+	switch app.Status.AppState.State {
+	case v1alpha1.NewState:
+		return true
+	case v1alpha1.SubmittedState, v1alpha1.RunningState:
 		return false
-	}
+	case v1alpha1.CompletedState:
+		if app.Spec.RestartPolicy.Type == v1alpha1.Always {
+			return true
+		}
+	case v1alpha1.FailedState:
+		shouldRetry := false
+		if app.Spec.RestartPolicy.Type == v1alpha1.Always {
+			shouldRetry = true
+		} else if app.Spec.RestartPolicy.Type == v1alpha1.OnFailure {
+			// We Retry if we haven't hit the retry limit.
+			if app.Spec.RestartPolicy.OnFailureRetries != nil && app.Status.Attempts < (*app.Spec.RestartPolicy.OnFailureRetries)+1 {
+				shouldRetry = true
```

Comment:
seems repetitive logic, can we simplify?

Reply:
there are  a bunch of different config values used for different states.....let's chat tomorrow.